crates/sui-graphql-rpc/src/types/move_package.rs

```diff
@@ -430,6 +517,64 @@ impl MovePackage {
     }
 }
 
+#[async_trait::async_trait]
+impl Loader<PackageVersionKey> for Db {
+    type Value = SuiAddress;
+    type Error = Error;
+
+    async fn load(
+        &self,
+        keys: &[PackageVersionKey],
+    ) -> Result<HashMap<PackageVersionKey, SuiAddress>, Error> {
+        use packages::dsl;
+        let other = diesel::alias!(packages as other);
+
+        let id_versions: BTreeSet<_> = keys
+            .iter()
+            .map(|k| (k.address.into_vec(), k.version as i64))
+            .collect();
+
+        let stored_packages: Vec<(Vec<u8>, i64, Vec<u8>)> = self
+            .execute(move |conn| {
+                conn.results(|| {
+                    let mut query = dsl::packages
+                        .inner_join(other.on(dsl::original_id.eq(other.field(dsl::original_id))))
+                        .select((
+                            dsl::package_id,
+                            other.field(dsl::package_version),
+                            other.field(dsl::package_id),
+                        ))
+                        .into_boxed();
+
+                    for (id, version) in id_versions.iter().cloned() {
+                        query = query.or_filter(
+                            dsl::package_id
+                                .eq(id)
+                                .and(other.field(dsl::package_version).eq(version)),
+                        );
+                    }
+
+                    query
+                })
+            })
+            .await
```

Comment:
iirc, we do the self join because the `package_id` and `package_version` provided by some client may not point to the true tuple of `(package_id, package_version)`, hence we select `((dsl::package_id, other::package_version), other::package_id)` for the true `package_id` that corresponds to the `package_version` for the given `package_id`. The self join on `original_id` links these packages together

Reply:
Originally, I tried to rely only on `packages`, but unfortunately it doesn't work because we can convert any `MovePackage` into an `Object` and that requires having all the fields of the `Object` around. It also means that we need to provide the same consistency guarantees for packages as we do for objects in case they get converted.

I originally did have a plan to use `packages` for this -- I think you only need to store the previous transaction digest to make it work (you can create an instance of `StoredHistoryObject` from an instance of `StoredPackage` that has all those fields), but the consistency requirements is what ruins things.  Instead, I'm usin `packages` analogously to `objects_version` here.

I also had a plan to re-use the package resolver's cache for loading packages in this way (or vice versa), but this realisation makes that impossible too.